File: 24-Swap-Nodes-in-Pairs.js
/**
 * Definition for singly-linked list.
 */
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next);
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function (head) {
    let dummy = new ListNode(0);
    dummy.next = head;
    let curr = dummy;

    while (curr.next && curr.next.next) {
        let first = curr.next,
            second = curr.next.next;

        // swapping
        first.next = second.next;
        second.next = first;
        curr.next = second;

        // move curr ptr two nodes ahead
        curr = curr.next.next;
    }

    return dummy.next;
};

File: 24-Swap-Nodes-in-Pairs.py
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        curr, prev = head, dummy

        while curr and curr.next:
            next = curr.next.next
            second = curr.next

            second.next = curr
            curr.next = next
            prev.next = second

            prev = curr
            curr = next

        return dummy.next
